File: internal/search/llm.go
// Package search provides keyword-based and LLM-powered tool search with
// query caching. The keyword searcher scores tools by token overlap with the query,
// while the optional Gemini-powered searcher adds semantic understanding of
// natural language queries.
//
// See https://hypercall.xyz for more information.

package search

import (
	"context"
	"fmt"

	"github.com/hypercall-public/mcpzip/internal/types"
)

// GeminiSearcher is a stub for LLM-based tool search re-ranking via Gemini.
// The real implementation will use the google/generative-ai-go SDK.
type GeminiSearcher struct {
	apiKey string
	model  string
}

// NewGeminiSearcher creates a new stub GeminiSearcher.
func NewGeminiSearcher(apiKey, model string) *GeminiSearcher {
	return &GeminiSearcher{apiKey: apiKey, model: model}
}

// Search is a stub that returns an error indicating LLM search is not yet implemented.
func (g *GeminiSearcher) Search(_ context.Context, _ string, _ int) ([]types.SearchResult, error) {
	return nil, fmt.Errorf("LLM search not yet implemented")
}

File: src/search/llm.rs
use crate::error::McpzipError;
use crate::types::SearchResult;

/// Stub for LLM-based tool search re-ranking via Gemini.
pub struct GeminiSearcher {
    _api_key: String,
    _model: String,
}

impl GeminiSearcher {
    pub fn new(api_key: String, model: String) -> Self {
        Self {
            _api_key: api_key,
            _model: model,
        }
    }

    /// Stub that returns an error indicating LLM search is not yet implemented.
    pub async fn search(&self, _query: &str, _limit: usize) -> Result<Vec<SearchResult>, McpzipError> {
        Err(McpzipError::Transport("LLM search not yet implemented".into()))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_stub_returns_error() {
        let gs = GeminiSearcher::new("key".into(), "model".into());
        let result = gs.search("slack", 10).await;
        assert!(result.is_err());
    }
}
